### PR TITLE
[SYCL-MLIR][INLINE]: Add aggressive & ludicrous heuristics

### PR DESCRIPTION
**Pull request by etiotto**
This PR expands upon PR https://github.com/intel/llvm/pull/7576 by adding the "aggressive" and "ludicrous" heuristics. It also modifies the "simple" heuristic. The inline heuristics 'mode' is as follows:
  - alwaysinline: collect 'sycl.call' edges
     + inline callees that have the 'alwaysinline' attribute
  - simple: collect 'sycl.call' and 'sycl.constructor' edges
     + inline callees that have either the 'inlinehint' or the 'alwaysinline' attribute, and 
     + inline callees that are dead after inlining the last call, and
     + inline small callees (num instructions < small threshold)
  - aggressive: collect 'sycl.call', 'sycl.constructor', and 'sycl.method' call instructions
     + inline callees that have either the 'inlinehint' or the 'alwaysinline' attribute, and 
     + inline callees that are dead after inlining the last call, and
     + inline medium size callees (num instructions < medium threshold)
  - ludicrous:  collect all call instructions
     + inline callees that have either the 'inlinehint' or the 'alwaysinline' attribute, and 
     + inline callees that are dead after inlining the last call, and
     + inline large size callees (num instructions < large threshold)
     
Finally, the PR allows callees containing instructions from the vector dialect to be inlined.

Signed-off-by: Tiotto, Ettore <ettore.tiotto@intel.com>

### CODE REVIEW

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 1]
```diff
@@ -220,8 +260,8 @@ class InlineHeuristic {
   /// Returns true if the target is an ancestor of the call and false otherwise.
   static bool isRecursiveCall(const ResolvedCall &ResolvedCall);
 
-  /// Inlining mode (alwaysinline, simple, ...)
-  const sycl::InlineMode InlineMode;
+  /// Returns the number of operations in the target of the supplied call.
+  static int64_t computeCalleeSize(const ResolvedCall &ResolvedCall);
```
Comment (etiotto) [inline mlir-sycl/lib/Transforms/Inliner.cpp:264]: Note: this is similar to what the Loop unrolling MLIR transformation does to evaluate the size of an unrolled iteration. 
Estimating the size if a function based on the number of instructions is imprecise because some dialect instructions might expand to several instructions in a lower dialect. If needed, we can refine this rough count by a combination of (as an example):
   - excluding certain dialect instructions from the count
   - assigning weights to instructions (i.e. instr. known to expand to large instr. counts might have a large weight)
   - return MAX_LLONG if the callee contains instructions in a 'disallowed list'

File: mlir-sycl/test/Transforms/inliner.mlir [hunk 2]
```diff
@@ -150,10 +175,10 @@ gpu.func @gpu_func_callee() -> i32 attributes {passthrough = ["alwaysinline"]} {
 // INLINE-DAG:     %c1_i32 = arith.constant 1 : i32
 // INLINE-DAG:     %c2_i32 = arith.constant 2 : i32
 // INLINE-DAG:     %c3_i32 = arith.constant 3 : i32
-// INLINE-DAG:     %0 = sycl.call() {FunctionName = @callee_, MangledFunctionName = @callee, TypeName = @A} : () -> i32
-// INLINE-NEXT:    %1 = arith.addi %c2_i32, %0 : i32
-// INLINE-NEXT:    %2 = arith.addi %c1_i32, %1 : i32
-// INLINE-NEXT:    %3 = arith.addi %c3_i32, %2 : i32
+// INLINE-DAG:     %0 = sycl.call() {FunctionName = @private_callee_, MangledFunctionName = @private_callee, TypeName = @A} : () -> i32
```
Comment (whitneywhtsang) [inline mlir-sycl/test/Transforms/inliner.mlir:178]: Why is this changed?

File: mlir-sycl/include/mlir/Dialect/SYCL/Transforms/Passes.h [hunk 3]
```diff
@@ -21,7 +21,7 @@ namespace mlir {
 namespace sycl {
 
 /// Inline mode to attempt.
-enum InlineMode { AlwaysInline, Simple, Aggressive };
+enum InlineMode { AlwaysInline, Simple, Aggressive, Ludicrous };
```
Comment (whitneywhtsang) [inline mlir-sycl/include/mlir/Dialect/SYCL/Transforms/Passes.h:24]: Should InlineMode also impact default `MaxIterationCount`?

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 4]
```diff
@@ -46,6 +48,33 @@ static FunctionOpInterface getCalledFunction(const CallOpInterface &Call) {
   return nullptr;
 }
 
+static bool isSYCLMethod(Operation *Op) {
+  using namespace sycl;
+
+  // TODO: find a way to avoid having to modify this code is a new SYCL Method
```
[the remark this answers is not in the document]
Comment (victor-eds) [reply mlir-sycl/lib/Transforms/Inliner.cpp:54]: We already have `llvm::isa<mlir::sycl::SYCLMethodOpInterface>(Op)`

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 5]
```diff
@@ -652,7 +720,25 @@ bool Inliner::inlineCallsInSCC(Inliner &Inliner, CGUseList &UseList,
...
+                             SmallVectorImpl<ResolvedCall> &Calls) const {
+  auto Collect = [this](const CallGraphNode &CGN, const CallOpInterface &Call) {
+    if (CGN.isExternal())
+      return false;
+
+    // Select which call operations to collect based on heuristics.
+    switch (Heuristic.InlineMode) {
+    case sycl::InlineMode::Ludicrous:
+      return isa<sycl::SYCLCallOp>(Call) ||
+             isa<sycl::SYCLConstructorOp>(Call) || isSYCLMethod(Call);
```
Comment (victor-eds) [inline mlir-sycl/lib/Transforms/Inliner.cpp:732]:
```suggestion
      return isa<sycl::SYCLCallOp, sycl::SYCLConstructorOp, sycl::SYCLMethodOpInterface>(Call);
```
Same can be applied to all cases.

**Review by victor-eds — COMMENTED** [review]
Minor comments. Also, what is the reason of not inlining constructors in the `Simple` mode?

File: mlir-sycl/lib/Transforms/Inliner.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (etiotto) [reply mlir-sycl/lib/Transforms/Inliner.cpp:54]: That is what I was looking for, good.

File: mlir-sycl/include/mlir/Dialect/SYCL/Transforms/Passes.h (hunk 3 above)
Comment (etiotto) [reply mlir-sycl/include/mlir/Dialect/SYCL/Transforms/Passes.h:24]: I'm inclined to keep then 2 separated but I'm open to preferences. Is that your preference?

Comment (etiotto) [reply mlir-sycl/include/mlir/Dialect/SYCL/Transforms/Passes.h:24]: Discussed offline. Decided we want to set the MaxIterationCount depending on the InlineMode (can always revisit this decision later if we need too).

**Review by whitneywhtsang — APPROVED** [review]

**Review by victor-eds — APPROVED** [review]
